chain/chunks/src/lib.rs

```diff
@@ -178,53 +204,124 @@ impl SealsManager {
+                        if part_owner == chunk_producer || Some(part_owner) == self.me {
+                            continue;
+                        }
+                        accumulator.push(part_ord);
+                    }
+
+                    accumulator
+                };
+
+                let chosen = candidates
+                    .choose_multiple(
+                        &mut rand::thread_rng(),
+                        cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
+                    )
+                    .cloned()
+                    .collect::<HashSet<_>>();
+                let challenge = ActiveSealChallenge {
+                    part_ords: chosen,
+                    chunk_producer,
+                    sent: Utc::now(),
+                    height,
+                };
+
+                Ok(entry.insert(challenge))
             }
-            let chosen = candidates
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
-                )
-                .cloned()
-                .collect::<HashSet<_>>();
-            Seal { part_ords: chosen, chunk_producer, sent: Utc::now() }
-        }))
+        }
     }
 
     fn approve_chunk(&mut self, chunk_hash: &ChunkHash) {
-        let seal = self.seals.get_mut(chunk_hash).expect("seal should be already produced");
-        seal.part_ords.clear();
+        let seal =
+            self.active_challenges.remove(chunk_hash).expect("seal should be already produced");
+        Self::insert_previous_seal(&mut self.previous_seals, seal.height, chunk_hash.clone());
     }
 
-    fn track_seals(&mut self) {
-        let now = Utc::now();
-        for (chunk_hash, seal) in self.seals.iter_mut() {
-            if seal.part_ords.len() > NUM_PARTS_LEFT_IN_SEAL
-                && (now - seal.sent).num_milliseconds() > ACCEPTING_SEAL_PERIOD_MS
-            {
-                warn!(target: "client", "Couldn't reconstruct chunk {:?} from {:?}, I'm {:?}", chunk_hash, seal.chunk_producer, self.me);
-                self.dont_include_chunks_from.insert(seal.chunk_producer.clone());
-                seal.part_ords.clear();
+    fn insert_previous_seal(
+        previous_seals: &mut BTreeMap<BlockHeight, HashSet<ChunkHash>>,
+        height: BlockHeight,
+        chunk_hash: ChunkHash,
+    ) {
+        let hashes_at_height = previous_seals.entry(height).or_insert_with(HashSet::new);
+        hashes_at_height.insert(chunk_hash);
+    }
+
+    fn prune_previous_seals(&mut self, block_hash: &CryptoHash) {
+        if let Ok(gc_stop_height) = self.runtime_adapter.get_gc_stop_height(block_hash) {
+            if let Some(least_height) = self.previous_seals.keys().next() {
+                if least_height < &gc_stop_height {
+                    let remaining_seals = self.previous_seals.split_off(&gc_stop_height);
```

Comment:
We can use @bowenwang1996 idea to `split_off(least_height + 1)` here.

Reply:
@Kouprin please see https://github.com/nearprotocol/nearcore/pull/2702#discussion_r429470672 the discussion there is not done yet.

Reply:
What we're talking about?
I see no problem here except possibility of having too many Seals to remove after jumping gc_height. If it's not enough to clear height by height - in theoretical case that we received too many Chunks of same height and stored all of them as valid, which doesn't sound possible - we can manually remove 100 random (or first) hashes from `least_height`.

Reply:
> I see no problem here except possibility of having too many Seals to remove after jumping gc_height.

That's exactly what we were discussing. Notice that the first jump would mean more than 200k heights getting removed.

> If it's not enough to clear height by height - in theoretical case that we received too many Chunks of same height and stored all of them as valid, which doesn't sound possible - we can manually remove 100 random (or first) hashes from least_height.

Not sure why you think it is impossible. I don't see any protection against it.

Reply:
How many valid chunks can be produced per height? I assume each Chunk Producer can build no more than one valid Chunk. All other Chunks will not pass validation.

Reply:
I changed to the version with a fixed height horizon in https://github.com/nearprotocol/nearcore/pull/2702/commits/1b1b509156500d7da4d48e9ee4869388d562553e

I think your idea @Kouprin was to amortize the cost of getting rid of past seals by removing a couple in every call. But I worried that we could get into a situation where we were not removing fast enough. For example @bowenwang1996 indicated the first GC jump is bigger than the other ones, so we will still be cleaning up from the first jump when the second jump happens.

I think having a fixed height horizon should be ok so long as we set it large enough, and we can discuss how large that should be. But if I am misunderstanding something here then please let me know.

Reply:
@birchmd Epoch Changing is a rare event. It happens twice a day. However, tracking seals may be executed several times per second.

@birchmd We strictly rely on fact that Epoch Changing happens as rarely to keep us enough room to complete all tasks to stabilize the system. For example, if there is no Block produced while Epoch, the system will crash.